File: chrome/test/data/webui/glic/browser_tests/glic_drag_and_drop_browsertest.ts
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// cc_file_path: chrome/browser/glic/host/glic_drag_and_drop_browsertest.cc

import {ApiTestFixtureBase, assertEquals, testMain, waitFor} from './browser_test_base.js';

class GlicDragAndDropBrowserTest extends ApiTestFixtureBase {
  override async setUpTest() {
    await this.client.waitForFirstOpen();
  }

  async testDragAndDropFile() {
    const dragEnter = Promise.withResolvers<void>();
    const dragOver = Promise.withResolvers<void>();
    const nativeDrop = Promise.withResolvers<string>();

    window.addEventListener('dragenter', () => {
      dragEnter.resolve();
    }, {once: true});
    window.addEventListener('dragover', (e: DragEvent) => {
      e.preventDefault();
      dragOver.resolve();
    });
    window.addEventListener('drop', (e: DragEvent) => {
      e.preventDefault();
      const files = e.dataTransfer?.files;
      nativeDrop.resolve(files && files.length > 0 ? files[0]!.name : '');
    }, {once: true});

    await this.advanceToNextStep();

    await waitFor(dragEnter.promise, 40000, 'DragEnter never arrived');
    await waitFor(dragOver.promise, 40000, 'DragOver never arrived');
    await this.advanceToNextStep();

    const droppedData =
        await waitFor(nativeDrop.promise, 40000, 'Drop never arrived');

    assertEquals('test.txt', droppedData);
  }
}

testMain([GlicDragAndDropBrowserTest]);

File: chrome/browser/glic/host/glic_drag_and_drop_browsertest.cc
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/files/scoped_temp_dir.h"
#include "base/test/run_until.h"
#include "base/threading/thread_restrictions.h"
#include "chrome/browser/glic/service/glic_instance_impl.h"
#include "chrome/browser/glic/test_support/glic_drag_and_drop_test_base.h"
#include "chrome/test/base/drag_and_drop_test_utils.h"
#include "content/public/browser/render_widget_host.h"
#include "content/public/browser/render_widget_host_view.h"
#include "content/public/test/browser_test.h"
#include "ui/gfx/geometry/point.h"

namespace glic {
namespace {

class GlicDragAndDropBrowserTest : public GlicDragAndDropTestBase,
                                   public testing::WithParamInterface<bool> {
 public:
  GlicDragAndDropBrowserTest()
      : GlicDragAndDropTestBase(
            GlicTestJsPath("./glic_drag_and_drop_browsertest.js"),
            GetParam()) {}
};

IN_PROC_BROWSER_TEST_P(GlicDragAndDropBrowserTest, testDragAndDropFile) {
  ASSERT_OK_AND_ASSIGN(GlicInstanceImpl * glic_instance,
                       OpenGlicForActiveTab());
  Host* glic_host = &glic_instance->host();
  PrepareGuestForDrag(*glic_host);
  gfx::Point host_relative_point = GetGuestCenterInHost(*glic_host);

  base::ScopedAllowBlockingForTesting allow_blocking;
  base::ScopedTempDir temp_dir;
  base::FilePath test_file = CreateTestFile(
      temp_dir, "test.txt", "This is some test content for file drop.");

  drag_and_drop_test_utils::DragAndDropSimulator simulator(
      glic_host->webui_contents());
  ASSERT_TRUE(simulator.SimulateDragEnter(host_relative_point, test_file));

  ContinueJsTest();

  ASSERT_TRUE(base::test::RunUntil([&]() {
    return glic_host->GetGuestMainFrame()->GetRenderWidgetHost()->GetView() &&
           !glic_host->GetGuestMainFrame()
                ->GetRenderWidgetHost()
                ->GetView()
                ->GetViewBounds()
                .IsEmpty();
  }));
  host_relative_point = GetGuestCenterInHost(*glic_host);
  ASSERT_TRUE(simulator.SimulateDrop(host_relative_point));

  ContinueJsTest();
}

INSTANTIATE_TEST_SUITE_P(All,
                         GlicDragAndDropBrowserTest,
                         testing::Bool(),
                         [](const testing::TestParamInfo<bool>& info) {
                           return info.param ? "NoWebview" : "Webview";
                         });

}  // namespace
}  // namespace glic
